File: plugins/runtime-asset-tracker/scripts/runtime-asset-ledger.mjs
#!/usr/bin/env node
import { appendFileSync, closeSync, existsSync, mkdirSync, openSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { homedir, hostname, platform } from "node:os";
import { dirname, join, resolve } from "node:path";
import { randomUUID } from "node:crypto";
import { spawn, spawnSync } from "node:child_process";
import readline from "node:readline";

const TRACKED_ATTRIBUTE_KEYS = new Set([
  "name",
  "image",
  "container",
  "driver",
  "network",
  "scope",
  "type",
  "com.docker.compose.project",
  "com.docker.compose.service",
  "com.docker.compose.volume",
  "com.docker.compose.network",
  "org.opencontainers.image.revision",
]);

function parseArgs(values) {
  const out = { _: [], detail: [] };
  for (let index = 0; index < values.length; index += 1) {
    const value = values[index];
    if (!value.startsWith("--")) {
      out._.push(value);
      continue;
    }
    const key = value.slice(2);
    const next = values[index + 1];
    if (next === undefined || next.startsWith("--")) {
      out[key] = true;
      continue;
    }
    index += 1;
    if (key === "detail") out.detail.push(next);
    else out[key] = next;
  }
  return out;
}

function clean(value, max = 4096) {
  if (value === undefined || value === null || value === "") return undefined;
  return String(value).replace(/[\r\n\0]/g, " ").slice(0, max);
}

function defaultLedgerFile() {
  if (process.env.RUNTIME_ASSET_LEDGER_FILE) return resolve(process.env.RUNTIME_ASSET_LEDGER_FILE);
  if (platform() === "win32") {
    const root = process.env.LOCALAPPDATA || join(homedir(), "AppData", "Local");
    return join(root, "RuntimeAssetTracker", "events.jsonl");
  }
  const root = process.env.XDG_STATE_HOME || join(homedir(), ".local", "state");
  return join(root, "runtime-asset-tracker", "events.jsonl");
}

function commandOutput(command, args, options = {}) {
  const result = spawnSync(command, args, { encoding: "utf8", windowsHide: true, ...options });
  if (result.status !== 0) return "";
  return String(result.stdout || "").trim();
}

function currentGitSha(cwd = process.cwd()) {
  return clean(process.env.RUNTIME_ASSET_GIT_SHA || process.env.TARGET_COMMIT || commandOutput("git", ["rev-parse", "HEAD"], { cwd }), 64) || "unknown";
}

function detailObject(entries) {
  const out = {};
  for (const entry of entries || []) {
    const separator = entry.indexOf("=");
    if (separator < 1) continue;
    const key = clean(entry.slice(0, separator), 128);
    const value = clean(entry.slice(separator + 1));
    if (key && value !== undefined) out[key] = value;
  }
  return out;
}

function context(args = {}) {
  return {
    project: clean(args.project || process.env.RUNTIME_ASSET_PROJECT, 128) || "unknown",
    environment: clean(args.environment || process.env.RUNTIME_ASSET_ENVIRONMENT, 64) || "unknown",
    release: clean(args.release || process.env.RUNTIME_ASSET_RELEASE, 256) || "unknown",
    gitSha: clean(args["git-sha"] || process.env.RUNTIME_ASSET_GIT_SHA, 64) || currentGitSha(),
    owner: clean(args.owner || process.env.RUNTIME_ASSET_OWNER, 128) || "unknown",
  };
}

function appendEvent(args, payload = {}) {
  const ledgerFile = resolve(args.ledger || defaultLedgerFile());
  mkdirSync(dirname(ledgerFile), { recursive: true });
  const event = {
    schemaVersion: 1,
    eventId: randomUUID(),
    occurredAt: new Date().toISOString(),
    event: clean(args.event, 128) || "runtime.unknown",
    host: hostname(),
    ...context(args),
    ...payload,
  };
  const details = detailObject(args.detail);
  if (Object.keys(details).length > 0) event.details = details;
  appendFileSync(ledgerFile, `${JSON.stringify(event)}\n`, { encoding: "utf8", mode: 0o600 });
  return { ledgerFile, event };
}

function safeLabels(labels) {
  const out = {};
  for (const [key, value] of Object.entries(labels || {})) {
    if (key.startsWith("com.codex.runtime.") || key === "org.opencontainers.image.revision" || key === "org.opencontainers.image.source") {
      out[key] = clean(value, 1024);
    }
  }
  return out;
}

function dockerJson(args) {
  const output = commandOutput("docker", args);
  if (!output) throw new Error(`Docker command failed: docker ${args.join(" ")}`);
  return JSON.parse(output);
}

function recordImage(args) {
  if (!args.image) throw new Error("image command requires --image");
  const inspected = dockerJson(["image", "inspect", args.image]);
  const image = inspected[0];
  return appendEvent(args, {
    asset: {
      type: "image",
      id: clean(image?.Id, 256),
      reference: clean(args.image, 512),
      repoTags: Array.isArray(image?.RepoTags) ? image.RepoTags.map((value) => clean(value, 512)) : [],
      repoDigests: Array.isArray(image?.RepoDigests) ? image.RepoDigests.map((value) => clean(value, 512)) : [],
      createdAt: clean(image?.Created, 128),
      sizeBytes: Number.isFinite(image?.Size) ? image.Size : undefined,
      labels: safeLabels(image?.Config?.Labels),
      service: clean(args.service, 128),
    },
  });
}

function jsonLines(commandArgs) {
  const output = commandOutput("docker", commandArgs);
  if (!output) return [];
  return output.split(/\r?\n/).filter(Boolean).flatMap((line) => {
    try { return [JSON.parse(line)]; } catch { return []; }
  });
}

function snapshot(args) {
  const containers = jsonLines(["ps", "-a", "--format", "{{json .}}"])
    .map((item) => ({ id: item.ID, name: item.Names, image: item.Image, state: item.State, status: item.Status }));
  const images = jsonLines(["image", "ls", "--no-trunc", "--format", "{{json .}}"])
    .map((item) => ({ id: item.ID, repository: item.Repository, tag: item.Tag, createdAt: item.CreatedAt, size: item.Size }));
  const volumes = jsonLines(["volume", "ls", "--format", "{{json .}}"])
    .map((item) => ({ name: item.Name, driver: item.Driver, scope: item.Scope }));
  const networks = jsonLines(["network", "ls", "--no-trunc", "--format", "{{json .}}"])
    .map((item) => ({ id: item.ID, name: item.Name, driver: item.Driver, scope: item.Scope }));
  return appendEvent({ ...args, event: args.event || "inventory.snapshot" }, {
    inventory: { containers, images, volumes, networks },
  });
}

function filteredAttributes(attributes) {
  const out = {};
  for (const [key, value] of Object.entries(attributes || {})) {
    if (TRACKED_ATTRIBUTE_KEYS.has(key) || key.startsWith("com.codex.runtime.")) out[key] = clean(value, 1024);
  }
  return out;
}

function pidAlive(pid) {
  try { process.kill(pid, 0); return true; } catch { return false; }
}

function acquireWatcherLock(ledgerFile) {
  const lockFile = `${ledgerFile}.watch.lock`;
  mkdirSync(dirname(lockFile), { recursive: true });
  if (existsSync(lockFile)) {
    const oldPid = Number(readFileSync(lockFile, "utf8").trim());
    if (Number.isInteger(oldPid) && oldPid > 0 && pidAlive(oldPid)) return null;
    rmSync(lockFile, { force: true });
  }
  const fd = openSync(lockFile, "wx", 0o600);
  writeFileSync(fd, `${process.pid}\n`, "utf8");
  closeSync(fd);
  return lockFile;
}

async function watch(args) {
  const ledgerFile = resolve(args.ledger || defaultLedgerFile());
  const lockFile = acquireWatcherLock(ledgerFile);
  if (!lockFile) return;
  let stopping = false;
  const stop = () => { stopping = true; };
  process.on("SIGINT", stop);
  process.on("SIGTERM", stop);
  appendEvent({ ...args, ledger: ledgerFile, event: "observer.started" }, { observerPid: process.pid });
  try { snapshot({ ...args, ledger: ledgerFile, event: "inventory.bootstrap" }); } catch (error) {
    appendEvent({ ...args, ledger: ledgerFile, event: "inventory.bootstrap.failed", detail: [`error=${error.message}`] });
  }
  try {
    while (!stopping) {
      const child = spawn("docker", ["events", "--format", "{{json .}}"], { stdio: ["ignore", "pipe", "pipe"], windowsHide: true });
      const lines = readline.createInterface({ input: child.stdout, crlfDelay: Infinity });
      for await (const line of lines) {
        if (stopping) break;
        try {
          const raw = JSON.parse(line);
          const type = clean(raw.Type || raw.type, 64) || "unknown";
          const rawAction = String(raw.Action || raw.action || raw.status || "unknown");
          const action = clean(rawAction.split(":", 1)[0].trim(), 64) || "unknown";
          const attributes = filteredAttributes(raw.Actor?.Attributes);
          appendEvent({ ...args, ledger: ledgerFile, event: `docker.${type}.${action}` }, {
            project: clean(attributes["com.codex.runtime.project"] || attributes["com.docker.compose.project"], 128) || context(args).project,
            environment: clean(attributes["com.codex.runtime.environment"], 64) || context(args).environment,
            release: clean(attributes["com.codex.runtime.release"], 256) || context(args).release,
            gitSha: clean(attributes["com.codex.runtime.git-sha"] || attributes["org.opencontainers.image.revision"], 64) || context(args).gitSha,
            owner: clean(attributes["com.codex.runtime.owner"], 128) || context(args).owner,
            asset: {
              type,
              id: clean(raw.Actor?.ID || raw.id, 256),
              attributes,
            },
            dockerTimeNano: clean(raw.timeNano, 64),
          });
        } catch {
          // Ignore malformed daemon lines. No raw line is persisted because it may contain unknown attributes.
        }
      }
      await new Promise((resolveDelay) => setTimeout(resolveDelay, 3000));
    }
  } finally {
    rmSync(lockFile, { force: true });
    appendEvent({ ...args, ledger: ledgerFile, event: "observer.stopped" }, { observerPid: process.pid });
  }
}

function usage() {
  console.error("Usage: runtime-asset-ledger.mjs <record|image|snapshot|watch> [--event name] [--ledger path] [--project key] [--environment name]");
}

const [command, ...rest] = process.argv.slice(2);
const args = parseArgs(rest);

try {
  if (command === "record") appendEvent(args, {
    asset: args["asset-type"] || args["asset-id"] ? {
      type: clean(args["asset-type"], 64),
      id: clean(args["asset-id"], 256),
      service: clean(args.service, 128),
    } : undefined,
    status: clean(args.status, 64),
  });
  else if (command === "image") recordImage(args);
  else if (command === "snapshot") snapshot(args);
  else if (command === "watch") await watch(args);
  else { usage(); process.exitCode = 2; }
} catch (error) {
  console.error(`runtime-asset-ledger: ${error.message}`);
  process.exitCode = 1;
}

File: plugins/runtime-asset-tracker/scripts/runtime-asset-ledger.py
#!/usr/bin/env python3
"""Standard-library fallback for the Runtime Asset Tracker ledger."""

from __future__ import annotations

import argparse
import json
import os
import platform
import signal
import socket
import subprocess
import sys
import time
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


TRACKED_ATTRIBUTE_KEYS = {
    "name",
    "image",
    "container",
    "driver",
    "network",
    "scope",
    "type",
    "com.docker.compose.project",
    "com.docker.compose.service",
    "com.docker.compose.volume",
    "com.docker.compose.network",
    "org.opencontainers.image.revision",
}


def clean(value: Any, maximum: int = 4096) -> str | None:
    if value is None or value == "":
        return None
    return str(value).replace("\r", " ").replace("\n", " ").replace("\0", " ")[:maximum]


def compact(value: Any) -> Any:
    if isinstance(value, dict):
        return {key: compact(item) for key, item in value.items() if item is not None}
    if isinstance(value, list):
        return [compact(item) for item in value if item is not None]
    return value


def default_ledger_file() -> Path:
    configured = os.environ.get("RUNTIME_ASSET_LEDGER_FILE")
    if configured:
        return Path(configured).expanduser().resolve()
    if os.name == "nt":
        root = Path(os.environ.get("LOCALAPPDATA", Path.home() / "AppData" / "Local"))
        return root / "RuntimeAssetTracker" / "events.jsonl"
    root = Path(os.environ.get("XDG_STATE_HOME", Path.home() / ".local" / "state"))
    return root / "runtime-asset-tracker" / "events.jsonl"


def command_output(command: list[str], cwd: str | None = None) -> str:
    result = subprocess.run(command, cwd=cwd, text=True, capture_output=True, check=False)
    return result.stdout.strip() if result.returncode == 0 else ""


def current_git_sha() -> str:
    return clean(
        os.environ.get("RUNTIME_ASSET_GIT_SHA")
        or os.environ.get("TARGET_COMMIT")
        or command_output(["git", "rev-parse", "HEAD"]),
        64,
    ) or "unknown"


def context(args: argparse.Namespace) -> dict[str, str]:
    return {
        "project": clean(args.project or os.environ.get("RUNTIME_ASSET_PROJECT"), 128) or "unknown",
        "environment": clean(args.environment or os.environ.get("RUNTIME_ASSET_ENVIRONMENT"), 64) or "unknown",
        "release": clean(args.release or os.environ.get("RUNTIME_ASSET_RELEASE"), 256) or "unknown",
        "gitSha": clean(args.git_sha or os.environ.get("RUNTIME_ASSET_GIT_SHA"), 64) or current_git_sha(),
        "owner": clean(args.owner or os.environ.get("RUNTIME_ASSET_OWNER"), 128) or "unknown",
    }


def details(entries: list[str]) -> dict[str, str]:
    result: dict[str, str] = {}
    for entry in entries:
        key, separator, value = entry.partition("=")
        safe_key = clean(key, 128)
        safe_value = clean(value)
        if separator and safe_key and safe_value is not None:
            result[safe_key] = safe_value
    return result


def append_event(args: argparse.Namespace, payload: dict[str, Any] | None = None) -> dict[str, Any]:
    ledger_file = Path(args.ledger).expanduser().resolve() if args.ledger else default_ledger_file()
    ledger_file.parent.mkdir(parents=True, exist_ok=True)
    event: dict[str, Any] = {
        "schemaVersion": 1,
        "eventId": str(uuid.uuid4()),
        "occurredAt": datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z"),
        "event": clean(args.event, 128) or "runtime.unknown",
        "host": socket.gethostname(),
        **context(args),
        **(payload or {}),
    }
    event_details = details(args.detail)
    if event_details:
        event["details"] = event_details
    event = compact(event)
    with ledger_file.open("a", encoding="utf-8") as handle:
        if os.name != "nt":
            os.chmod(ledger_file, 0o600)
        handle.write(json.dumps(event, ensure_ascii=False, separators=(",", ":")) + "\n")
    return {"ledgerFile": str(ledger_file), "event": event}


def safe_labels(labels: dict[str, Any] | None) -> dict[str, str]:
    result: dict[str, str] = {}
    for key, value in (labels or {}).items():
        if key.startswith("com.codex.runtime.") or key in {
            "org.opencontainers.image.revision",
            "org.opencontainers.image.source",
        }:
            safe_value = clean(value, 1024)
            if safe_value is not None:
                result[key] = safe_value
    return result


def docker_json(arguments: list[str]) -> Any:
    output = command_output(["docker", *arguments])
    if not output:
        raise RuntimeError(f"Docker command failed: docker {' '.join(arguments)}")
    return json.loads(output)


def json_lines(arguments: list[str]) -> list[dict[str, Any]]:
    output = command_output(["docker", *arguments])
    result: list[dict[str, Any]] = []
    for line in output.splitlines():
        try:
            item = json.loads(line)
            if isinstance(item, dict):
                result.append(item)
        except json.JSONDecodeError:
            continue
    return result


def record_image(args: argparse.Namespace) -> None:
    if not args.image:
        raise ValueError("image command requires --image")
    inspected = docker_json(["image", "inspect", args.image])
    image = inspected[0]
    append_event(args, {
        "asset": {
            "type": "image",
            "id": clean(image.get("Id"), 256),
            "reference": clean(args.image, 512),
            "repoTags": [clean(value, 512) for value in image.get("RepoTags") or []],
            "repoDigests": [clean(value, 512) for value in image.get("RepoDigests") or []],
            "createdAt": clean(image.get("Created"), 128),
            "sizeBytes": image.get("Size") if isinstance(image.get("Size"), (int, float)) else None,
            "labels": safe_labels((image.get("Config") or {}).get("Labels")),
            "service": clean(args.service, 128),
        }
    })


def snapshot(args: argparse.Namespace) -> None:
    containers = [
        {"id": item.get("ID"), "name": item.get("Names"), "image": item.get("Image"), "state": item.get("State"), "status": item.get("Status")}
        for item in json_lines(["ps", "-a", "--format", "{{json .}}"])
    ]
    images = [
        {"id": item.get("ID"), "repository": item.get("Repository"), "tag": item.get("Tag"), "createdAt": item.get("CreatedAt"), "size": item.get("Size")}
        for item in json_lines(["image", "ls", "--no-trunc", "--format", "{{json .}}"])
    ]
    volumes = [
        {"name": item.get("Name"), "driver": item.get("Driver"), "scope": item.get("Scope")}
        for item in json_lines(["volume", "ls", "--format", "{{json .}}"])
    ]
    networks = [
        {"id": item.get("ID"), "name": item.get("Name"), "driver": item.get("Driver"), "scope": item.get("Scope")}
        for item in json_lines(["network", "ls", "--no-trunc", "--format", "{{json .}}"])
    ]
    original_event = args.event
    args.event = args.event or "inventory.snapshot"
    append_event(args, {"inventory": {"containers": containers, "images": images, "volumes": volumes, "networks": networks}})
    args.event = original_event


def filtered_attributes(attributes: dict[str, Any] | None) -> dict[str, str]:
    result: dict[str, str] = {}
    for key, value in (attributes or {}).items():
        if key in TRACKED_ATTRIBUTE_KEYS or key.startswith("com.codex.runtime."):
            safe_value = clean(value, 1024)
            if safe_value is not None:
                result[key] = safe_value
    return result


def watch(args: argparse.Namespace) -> None:
    if platform.system() == "Windows":
        raise RuntimeError("Python watch fallback requires Linux; use the Node.js watcher on Windows")
    import fcntl

    ledger_file = Path(args.ledger).expanduser().resolve() if args.ledger else default_ledger_file()
    ledger_file.parent.mkdir(parents=True, exist_ok=True)
    lock_path = Path(f"{ledger_file}.watch.lock")
    lock_handle = lock_path.open("a+", encoding="utf-8")
    try:
        fcntl.flock(lock_handle.fileno(), fcntl.LOCK_EX | fcntl.LOCK_NB)
    except BlockingIOError:
        lock_handle.close()
        return
    lock_handle.seek(0)
    lock_handle.truncate()
    lock_handle.write(f"{os.getpid()}\n")
    lock_handle.flush()

    stopping = False
    child: subprocess.Popen[str] | None = None

    def stop(_signum: int, _frame: Any) -> None:
        nonlocal stopping, child
        stopping = True
        if child and child.poll() is None:
            child.terminate()

    signal.signal(signal.SIGINT, stop)
    signal.signal(signal.SIGTERM, stop)
    args.ledger = str(ledger_file)
    original_event = args.event
    args.event = "observer.started"
    append_event(args, {"observerPid": os.getpid()})
    try:
        args.event = "inventory.bootstrap"
        snapshot(args)
    except Exception as error:  # Docker may be temporarily unavailable during boot.
        args.event = "inventory.bootstrap.failed"
        args.detail = [*args.detail, f"error={error}"]
        append_event(args)
    try:
        while not stopping:
            child = subprocess.Popen(
                ["docker", "events", "--format", "{{json .}}"],
                text=True,
                stdout=subprocess.PIPE,
                stderr=subprocess.DEVNULL,
            )
            assert child.stdout is not None
            for line in child.stdout:
                if stopping:
                    break
                try:
                    raw = json.loads(line)
                    asset_type = clean(raw.get("Type") or raw.get("type"), 64) or "unknown"
                    raw_action = str(raw.get("Action") or raw.get("action") or raw.get("status") or "unknown")
                    action = clean(raw_action.split(":", 1)[0].strip(), 64) or "unknown"
                    actor = raw.get("Actor") or {}
                    attributes = filtered_attributes(actor.get("Attributes"))
                    fallback = context(args)
                    args.event = f"docker.{asset_type}.{action}"
                    append_event(args, {
                        "project": clean(attributes.get("com.codex.runtime.project") or attributes.get("com.docker.compose.project"), 128) or fallback["project"],
                        "environment": clean(attributes.get("com.codex.runtime.environment"), 64) or fallback["environment"],
                        "release": clean(attributes.get("com.codex.runtime.release"), 256) or fallback["release"],
                        "gitSha": clean(attributes.get("com.codex.runtime.git-sha") or attributes.get("org.opencontainers.image.revision"), 64) or fallback["gitSha"],
                        "owner": clean(attributes.get("com.codex.runtime.owner"), 128) or fallback["owner"],
                        "asset": {"type": asset_type, "id": clean(actor.get("ID") or raw.get("id"), 256), "attributes": attributes},
                        "dockerTimeNano": clean(raw.get("timeNano"), 64),
                    })
                except (json.JSONDecodeError, TypeError, ValueError):
                    continue
            if not stopping:
                time.sleep(3)
    finally:
        args.event = "observer.stopped"
        append_event(args, {"observerPid": os.getpid()})
        args.event = original_event
        fcntl.flock(lock_handle.fileno(), fcntl.LOCK_UN)
        lock_handle.close()
        lock_path.unlink(missing_ok=True)


def parser() -> argparse.ArgumentParser:
    command_parser = argparse.ArgumentParser(prog="runtime-asset-ledger.py")
    command_parser.add_argument("command", choices=("record", "image", "snapshot", "watch"))
    command_parser.add_argument("--event")
    command_parser.add_argument("--ledger")
    command_parser.add_argument("--project")
    command_parser.add_argument("--environment")
    command_parser.add_argument("--release")
    command_parser.add_argument("--git-sha", dest="git_sha")
    command_parser.add_argument("--owner")
    command_parser.add_argument("--detail", action="append", default=[])
    command_parser.add_argument("--asset-type")
    command_parser.add_argument("--asset-id")
    command_parser.add_argument("--service")
    command_parser.add_argument("--status")
    command_parser.add_argument("--image")
    return command_parser


def main() -> int:
    args = parser().parse_args()
    try:
        if args.command == "record":
            asset = None
            if args.asset_type or args.asset_id:
                asset = {"type": clean(args.asset_type, 64), "id": clean(args.asset_id, 256), "service": clean(args.service, 128)}
            append_event(args, {"asset": asset, "status": clean(args.status, 64)})
        elif args.command == "image":
            record_image(args)
        elif args.command == "snapshot":
            snapshot(args)
        else:
            watch(args)
        return 0
    except Exception as error:
        print(f"runtime-asset-ledger: {error}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
